network/gossip/libp2p/peerManager.go

```diff
@@ -0,0 +1,164 @@
+package libp2p
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/hashicorp/go-multierror"
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/model/flow/filter"
+)
+
+// Connector connects to peer and disconnects from peer using the underlying networking library
+type Connector interface {
+
+	// ConnectPeers connect to the given flow.Identities and returns a map of identifiers which failed.
+	// ConnectPeer implementation should be idempotent such that multiple calls to connect to the same peer should not
+	// return an error or create multiple connections
+	ConnectPeers(ctx context.Context, ids flow.IdentityList) map[flow.Identifier]error
+
+	// DisconnectPeers disconnect from the given flow.Identities and returns a map of identifiers which failed.
+	// DisconnectPeers implementation should be idempotent such that multiple calls to connect to the same peer should
+	// not return an error
+	DisconnectPeers(ctx context.Context, ids flow.IdentityList) map[flow.Identifier]error
+}
+
+// PeerUpdateInterval is how long the peer manager waits in between attempts to update peer connections
+var PeerUpdateInterval = 1 * time.Minute
+
+// PeerManager adds and removes connections to peers periodically and on request
+type PeerManager struct {
+	ctx         context.Context
+	logger      zerolog.Logger
+	idsProvider func() (flow.IdentityList, error) // callback to retrieve list of peers to connect to
+
+	peerRequestQ chan struct{}     // a channel to queue a peer update request
+	connector    Connector         // connector to connect or disconnect from peers
+	currentIDs   flow.IdentityList // a list of currentIDs this node should be connected to
+}
+
+// NewPeerManager creates a new peer manager which calls the idsProvider callback to get a list of peers to connect to
+// and it uses the connector to actually connect or disconnect from peers.
+func NewPeerManager(ctx context.Context, logger zerolog.Logger, idsProvider func() (flow.IdentityList, error), connector Connector) *PeerManager {
+	return &PeerManager{
+		ctx:          ctx,
+		logger:       logger,
+		idsProvider:  idsProvider,
+		connector:    connector,
+		peerRequestQ: make(chan struct{}, 1),
```

Comment:
Shouldn't this be `make(chan struct{})`, i.e., an unbuffered channel? otherwise, while an update is in progress (i.e., already requested), a new one is _buffered_, and will no longer be a `no-op`:

> `// RequestPeerUpdate requests an update to the peer connections of this node.
// If a peer update has already been requested (either as a periodic request or an on-demand request) and is outstanding,`

Reply:
So there is `default` case in `RequestPeerUpdate` which will make sure the caller never blocks (https://github.com/onflow/flow-go/pull/78/files#diff-e0ff51e04cd65ba801a58cba4cbcea1275bfef8a56860e1e369b2a9f7733dc68R99).
The `make(chan struct{}, 1)` makes sure that any given time, there is only one outstanding request for update (beyond the one that may be running at that time).
I added a unit test to make sure that is indeed the case: https://github.com/onflow/flow-go/pull/78/files#diff-0c6db110aaecc667499723618d5f44f537a809837096cc995421bfb876f80655R173